### PR TITLE
Add overflow checks to parse_number/parse_hex/parse_oct

### PR DESCRIPTION
**Pull request by nhorman**
Test the next arithmetic operation to safely determine if adding the next digit in the passed property string will overflow

Also, noted a bug in the parse_hex code.  When parsing non-digit characters (i.e. a-f and A-F), we do a tolower conversion (which is fine), and then subtract 'a' to get the hex value from the ascii (which is definitely wrong).  We should subtract 'W' to convert tolower converted hex digits in the range a-f to their hex value counterparts

Add tests to test_property_parse_error to ensure overflow checks work

##### Checklist
- [x] tests are added or updated

### CODE REVIEW

File: crypto/property/property_parse.c [hunk 1]
```diff
@@ -117,15 +126,27 @@ static int parse_hex(const char *t[], OSSL_PROPERTY_DEFINITION *res)
...
-        return 0;
     do {
-        v <<= 4;
         if (ossl_isdigit(*s))
-            v += *s - '0';
-        else
-            v += ossl_tolower(*s) - 'a';
+            sval = *s - '0';
+        else if (ossl_isxdigit(*s))
+            sval = ossl_tolower(*s) - 'W';
```
Comment (t-j-h) [inline crypto/property/property_parse.c:135]: That really isn't clear code - it should be subtract 'a' **and add 10** logically.
This has nothing to do with 'W' and with no comment about it that really doesn't work well.

Note this also won't work on EBCDIC systems ...

Comment (nhorman) [reply crypto/property/property_parse.c:135]: ascii 'W' is the equivalent of subtracting ascii 'a' and adding 10, but if we support EBCDIC encoding I can make the appropriate change

Comment (t-j-h) [reply crypto/property/property_parse.c]: I know what the code is doing - but writing "W" *obfuscates* the logic - so both the obfuscation and the non-EBCDIC issue apply - i.e. even if you ignore EBCDIC that code should be changed IMHO.

**Review by t8m — CHANGES_REQUESTED** [review]

**Review by t8m — APPROVED** [review]

File: test/property_test.c [hunk 2]
```diff
@@ -198,6 +199,9 @@ static const struct {
     { 1, "a=2, n=012345678" },  /* Bad octal digit */
     { 0, "n=0x28FG, a=3" },     /* Bad hex digit */
     { 0, "n=145d, a=2" },       /* Bad decimal digit */
+    { 0, "n=0xffffffffffffffff, a=3" },      /* Hex overflow */
+    { 0, "n=9223372037000000000d, a=2" },    /* Decimal overflow */
```
Comment (tom-cosgrove-arm) [inline test/property_test.c:203]: Ideally for these tests we'd have positive (should pass) tests for a value of `INT64_MAX` in dec, hex and octal, as well as negative (should fail) tests for a value of `INT64_MAX + 1` in dec, hex and octal.

**Review by levitte — APPROVED** [review]

File: test/property_test.c (hunk 2 above)
Comment (nhorman) [reply test/property_test.c]: modified tests such that negative tests use INT64_MAX+1 in decimal, hex and octal, and added positive tests for hex/decimal/octal using INT64_MAX

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@levitte please reconfirm

**Comment by richsalz** [comment]
Subtracting 'W' to convert [a-f] to 10-15 is a gross hack. You shoujld be doing ```- 'a' + 10```
